### PR TITLE
Pr cmsis dap rebase

### PR DESCRIPTION
**Pull request by maxd-nordic**

### CODE REVIEW

File: subsys/dap/Kconfig [hunk 1]
```diff
@@ -17,6 +17,13 @@ config CMSIS_DAP_PACKET_COUNT
 	help
 	  Maximum packet buffers for request and response data.
 
+config CMSIS_DAP_PACKET_SIZE
+	int "Maximum packet size for request and response data."
```
Comment (tmon-nordic) [inline subsys/dap/Kconfig:21]: This Kconfig has to be removed because this value is used to set wMaxPacketSize field in CONFIGURATION descriptor.

The endpoints are bulk, therefore the only valid options (Universal Serial Bus Specification Revision 2.0 5.8.3 Bulk Transfer Packet Size Constraints) are:
- 8, 16, 32 or 64 on Full-Speed device (or High-Speed capable device operating at Full-Speed)
- 512 on High-Speed device operating at High-Speed

Legacy USB stack (i.e. not device-next) is not really suitable for properly handling High-Speed (the problem is with High-Speed device operating at Full-Speed). The next stack is better on this.

While the 8/16/32/64 max packet size choice can seem arbitrary to the application software, there are cases where the USB device controller does not support all of the options. Due to this is better for the USB stack to handle the endpoint sizes.

For the old stack probably the best that can be done is probably just use 64 for Full-Speed and 512 when `CONFIG_USB_DC_HAS_HS_SUPPORT` is defined. This will exhibit the problem with High-Speed operating at Full-Speed, but fixing that in legacy stack seems infeasible.

**Review by tmon-nordic — CHANGES_REQUESTED** [review]

**Comment by maxd-nordic** [comment]
@jfischer-no We talked about the speed of probes that use zephyr_gpio instead of writing to registers directly.
I just compared erasing+flashing an nRF9160 with both configs and the results are good enough in my opinion: 49,342s (zephyr_gpio), 33,116s (direct register write).

File: subsys/dap/Kconfig (hunk 1 above)
Comment (maxd-nordic) [reply subsys/dap/Kconfig:21]: Would it be enough to set the default based `USB_DC_HAS_HS_SUPPORT` and add a help text?
Or do you think the risk of misconfiguration is too high?

**Comment by maxd-nordic** [comment]
Just tested on the Raspberry Pi Pico (rp_2040), which needs the new DAP command I implemented and it works fine!

File: subsys/dap/Kconfig [hunk 2]
```diff
@@ -24,6 +24,30 @@ config CMSIS_DAP_PACKET_SIZE
...
+	default ""
+
+config CMSIS_DAP_DEVICE_VENDOR
+	string "Target device vendor"
+	default ""
+
+config CMSIS_DAP_DEVICE_NAME
+	string "Target device name"
+	default ""
+
```
Comment (jfischer-no) [inline subsys/dap/Kconfig:50]: These options are more problematic than others in terms of multi-instance support.

File: subsys/dap/cmsis_dap.c [hunk 3]
```diff
@@ -59,30 +57,62 @@ static uint16_t dap_info(struct dap_context *const ctx,
 
 	switch (id) {
 	case DAP_ID_VENDOR:
-		LOG_DBG("ID_VENDOR unsupported");
+		LOG_DBG("ID_VENDOR");
+		memcpy(info, CONFIG_CMSIS_DAP_PROBE_VENDOR, sizeof(CONFIG_CMSIS_DAP_PROBE_VENDOR));
+		length = (uint8_t)sizeof(CONFIG_CMSIS_DAP_PROBE_VENDOR);
```
Comment (jfischer-no) [inline subsys/dap/cmsis_dap.c]: Why is there cast operator?

File: subsys/dap/cmsis_dap.c [hunk 4]
```diff
@@ -59,30 +57,62 @@ static uint16_t dap_info(struct dap_context *const ctx,
 
 	switch (id) {
 	case DAP_ID_VENDOR:
-		LOG_DBG("ID_VENDOR unsupported");
+		LOG_DBG("ID_VENDOR");
+		memcpy(info, CONFIG_CMSIS_DAP_PROBE_VENDOR, sizeof(CONFIG_CMSIS_DAP_PROBE_VENDOR));
```
Comment (jfischer-no) [inline subsys/dap/cmsis_dap.c]: There should be length check/limitations for info.

File: subsys/dap/cmsis_dap.c [hunk 5]
```diff
@@ -238,10 +270,16 @@ static uint16_t dap_swj_pins(struct dap_context *const ctx,
...
+	do
+	{
+		api->swdp_get_pins(ctx->swdp_dev, &state);
+		LOG_INF("select 0x%02x, value 0x%02x, wait %u, state 0x%02x",
+			select, value, wait, state);
+		if ((value & select) == (state & select)) {
+			LOG_DBG("swdp_get_pins succeeded before timeout");
+			break;
+		}
+	} while (k_uptime_ticks() - timout_ticks > 0);
```
Comment (jfischer-no) [inline subsys/dap/cmsis_dap.c:307]: If it implements wait then it should be part of separate commit.

File: subsys/dap/cmsis_dap.c [hunk 6]
```diff
@@ -570,6 +608,50 @@ static uint16_t dap_transfer(struct dap_context *const ctx,
...
+			response_data += num_bytes;
+		} else {
+			api->swdp_output_sequence(ctx->swdp_dev, num_cycles, request_data);
+			request_data += num_bytes;
+		}		
+	}
+
+	return response_data - response;
+}
+
```
Comment (jfischer-no) [inline subsys/dap/cmsis_dap.c:679]: Whatever it does, it should be part of separate commit.

File: subsys/dap/cmsis_dap.h [hunk 7]
```diff
@@ -35,20 +35,31 @@
...
 #define ID_DAP_SWO_BAUDRATE			0x19U
 #define ID_DAP_SWO_CONTROL			0x1AU
 #define ID_DAP_SWO_STATUS			0x1BU
 #define ID_DAP_SWO_DATA				0x1CU
 
+#define ID_DAP_UART_Transport			0x1FU
+#define ID_DAP_UART_Configure			0x20U
+#define ID_DAP_UART_Control			0x22U
+#define ID_DAP_UART_Status			0x23U
+#define ID_DAP_UART_Transfer			0x21U
```
Comment (jfischer-no) [inline subsys/dap/cmsis_dap.h]: Should be all caps.

File: subsys/dap/cmsis_dap.h [hunk 8]
```diff
@@ -60,30 +71,36 @@
 #define ID_DAP_INVALID				0xFFU
 
 /* DAP Status Code */
-#define DAP_OK					0U
+#define DAP_OK					0x00U
```
Comment (jfischer-no) [inline subsys/dap/cmsis_dap.h]: Stray changes

File: subsys/dap/cmsis_dap.h [hunk 9]
```diff
@@ -60,30 +71,36 @@
...
+#define DAP_ID_UART_TX_BUFFER_SIZE		0xFCU
 #define DAP_ID_SWO_BUFFER_SIZE			0xFDU
 #define DAP_ID_PACKET_COUNT			0xFEU
 #define DAP_ID_PACKET_SIZE			0xFFU
 
 /* DAP Host Status */
-#define DAP_DEBUGGER_CONNECTED			0U
-#define DAP_TARGET_RUNNING			1U
+#define DAP_DEBUGGER_CONNECTED			0x00U
+#define DAP_TARGET_RUNNING			0x01U
```
Comment (jfischer-no) [inline subsys/dap/cmsis_dap.h]: Stray changes.

File: subsys/dap/cmsis_dap.h [hunk 10]
```diff
@@ -60,30 +71,36 @@
...
 
 /* DAP Port */
-#define DAP_PORT_AUTODETECT			0U
-#define DAP_PORT_DISABLED			0U
-#define DAP_PORT_SWD				1U
-#define DAP_PORT_JTAG				2U
+#define DAP_PORT_AUTODETECT			0x00U
+#define DAP_PORT_DISABLED			0x00U
+#define DAP_PORT_SWD				0x01U
+#define DAP_PORT_JTAG				0x02U
```
Comment (jfischer-no) [inline subsys/dap/cmsis_dap.h]: Stray changes.

File: subsys/dap/cmsis_dap.h [hunk 11]
```diff
@@ -109,9 +126,9 @@
 #define DP_RESEND				0x08U
 #define DP_RDBUFF				0x0CU
 
-#define DAP_MBMSG_REGISTER_IFACE		0x0U
-#define DAP_MBMSG_FROM_IFACE			0x1U
-#define DAP_MBMSG_FROM_CONTROLLER		0x2U
+#define DAP_MBMSG_REGISTER_IFACE		0x00U
+#define DAP_MBMSG_FROM_IFACE			0x01U
+#define DAP_MBMSG_FROM_CONTROLLER		0x02U
```
Comment (jfischer-no) [inline subsys/dap/cmsis_dap.h]: Stray changes.

File: subsys/dap/cmsis_dap.h [hunk 12]
```diff
@@ -60,30 +71,36 @@
...
-#define DAP_OK					0U
+#define DAP_OK					0x00U
 #define DAP_ERROR				0xFFU
 
 /* DAP ID */
 #define DAP_ID_VENDOR				0x01U
 #define DAP_ID_PRODUCT				0x02U
 #define DAP_ID_SER_NUM				0x03U
-#define DAP_ID_FW_VER				0x04U
+#define DAP_ID_DAP_FW_VER			0x04U
```
Comment (jfischer-no) [inline subsys/dap/cmsis_dap.h]: Stray changes.

File: samples/subsys/dap/CMakeLists.txt [hunk 13]
```diff
@@ -1,9 +1,10 @@
-# Copyright (c) 2023 Nordic Semiconductor ASA
 # SPDX-License-Identifier: Apache-2.0
 
 cmake_minimum_required(VERSION 3.20.0)
 find_package(Zephyr REQUIRED HINTS $ENV{ZEPHYR_BASE})
-project(dap)
+project(webusb)
```
Comment (jfischer-no) [inline samples/subsys/dap/CMakeLists.txt]: It is not webusb.

File: samples/subsys/dap/CMakeLists.txt [hunk 14]
```diff
@@ -1,9 +1,10 @@
...
+
+target_include_directories(app PRIVATE ${ZEPHYR_BASE}/subsys/usb)
```
Comment (jfischer-no) [inline samples/subsys/dap/CMakeLists.txt]: What parts are you missing in public header?

File: samples/subsys/dap/boards/nrf52840dk_nrf52840.conf [hunk 15]
```diff
@@ -0,0 +1,4 @@
+CONFIG_SWDP_BITBANG_DRIVER_SEPARATE_IN_OUT=n
+
+CONFIG_CMSIS_DAP_DEVICE_VENDOR="Raspberry Pi"
+CONFIG_CMSIS_DAP_DEVICE_NAME="rp2040"
```
Comment (jfischer-no) [inline samples/subsys/dap/boards/nrf52840dk_nrf52840.conf]: Should not be Raspberry Pi

File: samples/subsys/dap/boards/nrf52840dk_nrf52840.overlay [hunk 16]
```diff
@@ -0,0 +1,26 @@
...
+		noe-gpios = <&arduino_header 11 GPIO_ACTIVE_HIGH>;	/* D5 */
+		reset-gpios = <&arduino_header 13 GPIO_ACTIVE_HIGH>;	/* D7 */
+		port-write-cycles = <2>;
+	};
+};
+
+&zephyr_udc0 {
+	cdc_acm_uart0 {
+		compatible = "zephyr,cdc-acm-uart";
+	};
```
Comment (jfischer-no) [inline samples/subsys/dap/boards/nrf52840dk_nrf52840.overlay]: What is cdc_acm_uart0 used for?

File: samples/subsys/dap/src/cmsis_dap_usb.c [hunk 17]
```diff
@@ -0,0 +1,409 @@
...
+	'3', 0x00, 'B', 0x00, '-', 0x00, '4', 0x00, '6', 0x00, '6', 0x00, \
+	'3', 0x00, '-', 0x00, 'A', 0x00, 'A', 0x00, '3', 0x00, '6', 0x00, \
+	'-', 0x00, '1', 0x00, 'A', 0x00, 'A', 0x00, 'E', 0x00, '4', 0x00, \
+	'6', 0x00, '4', 0x00, '6', 0x00, '3', 0x00, '7', 0x00, '7', 0x00, \
+	'6', 0x00, '}', 0x00, 0x00, 0x00, 0x00, 0x00
+
+#define COMPATIBLE_ID_WINUSB \
+	'W', 'I', 'N', 'U', 'S', 'B', 0x00, 0x00
+
+static struct msosv2_descriptor_t {
```
Comment (jfischer-no) [inline samples/subsys/dap/src/cmsis_dap_usb.c]: Please do not postfix "_t".

File: samples/subsys/dap/src/cmsis_dap_usb.c [hunk 18]
```diff
@@ -0,0 +1,409 @@
...
+	struct ms_os_20_descriptor_set_header_descriptor header;
+#if defined(CONFIG_USB_COMPOSITE_DEVICE)
+	struct ms_os_20_function_subset_header_descriptor cmsis_dap_v2_subset_header;
+#endif
```
Comment (jfischer-no) [inline samples/subsys/dap/src/cmsis_dap_usb.c]: Why do you think it should depend on COMPOSITE_DEVICE? I do not see an interface association descriptor.

File: subsys/dap/Kconfig (hunk 1 above)
Comment (tmon-nordic) [reply subsys/dap/Kconfig:21]: I don't think it should be Kconfig at all. There are already way too many Kconfigs that can give you essentially-broken-but-compiling code and this is going to be just another one.

File: samples/subsys/dap/src/cmsis_dap_usb.c [hunk 19]
```diff
@@ -0,0 +1,409 @@
...
+		*len = sizeof(msosv2_descriptor);
+
+		LOG_DBG("Get MS OS Descriptors v2");
+
+		return 0;
+	}
+	return -ENOTSUP;
+}
+
+void bulk_read_cb(uint8_t ep, int size, void *priv)
```
Comment (jfischer-no) [inline samples/subsys/dap/src/cmsis_dap_usb.c]: `static void bulk_read_cb(uint8_t ep, int size, void *priv)`

File: samples/subsys/dap/src/cmsis_dap_usb.c [hunk 20]
```diff
@@ -0,0 +1,409 @@
...
+{
+	struct net_buf *buf;
+	cfg = priv;
```
Comment (jfischer-no) [inline samples/subsys/dap/src/cmsis_dap_usb.c]: `struct usb_cfg_data *cfg = priv;`

File: samples/subsys/dap/src/cmsis_dap_usb.c [hunk 21]
```diff
@@ -0,0 +1,409 @@
...
+
+	LOG_DBG("cfg %p ep %x size %u", cfg, ep, size);
+
+	if (size <= 0) {
+		goto done;
+	}
+
+	buf = net_buf_alloc(&ep_out_pool, K_FOREVER);
+	memcpy(buf->data, rx_buf, CONFIG_CMSIS_DAP_PACKET_SIZE);
+	net_buf_add(buf, size);
```
Comment (jfischer-no) [inline samples/subsys/dap/src/cmsis_dap_usb.c]: `net_buf_add_mem(buf, rx_buf, MIN(size, CONFIG_CMSIS_DAP_PACKET_SIZE);`

File: samples/subsys/dap/src/cmsis_dap_usb.c [hunk 22]
```diff
@@ -0,0 +1,409 @@
...
+	k_fifo_put(&ep_out_queue, buf);
```
Comment (jfischer-no) [inline samples/subsys/dap/src/cmsis_dap_usb.c]: ep_out_queue is unlikely EP OUT queue but DAP subsystems input queue.

File: samples/subsys/dap/src/cmsis_dap_usb.c [hunk 23]
```diff
@@ -0,0 +1,409 @@
...
+static struct usb_ep_cfg_data dapusb_ep_data[] = {
+	{
+		.ep_cb	= usb_transfer_ep_callback,
+		.ep_addr = AUTO_EP_OUT
+	},
+	{
+		.ep_cb = usb_transfer_ep_callback,
+		.ep_addr = AUTO_EP_IN
+	}
+};
```
Comment (jfischer-no) [inline samples/subsys/dap/src/cmsis_dap_usb.c]: Please do not use  weird AUTO_EP_IN/AUTO_EP_OUT macros. Add
```
#define DAP_USB_EP_IN	0x81
#define DAP_USB_EP_OUT	0x01
```

File: samples/subsys/dap/src/cmsis_dap_usb.c [hunk 24]
```diff
@@ -0,0 +1,409 @@
...
+
+#define BULK_OUT_EP_IDX		0
+#define BULK_IN_EP_IDX		1
+
+static bool usb_bulk_ready;
```
Comment (jfischer-no) [inline samples/subsys/dap/src/cmsis_dap_usb.c]: Where is usb_bulk_ready used?

File: samples/subsys/dap/src/cmsis_dap_usb.c [hunk 25]
```diff
@@ -0,0 +1,409 @@
...
+		break;
+	case USB_DC_RESUME:
+		LOG_DBG("USB device resumed");
+		break;
+	case USB_DC_UNKNOWN:
+	default:
+		LOG_DBG("USB unknown state");
+		break;
+	}
+}
```
Comment (jfischer-no) [inline samples/subsys/dap/src/cmsis_dap_usb.c]: Please remove all the unused cases.

File: samples/subsys/dap/src/cmsis_dap_usb.c [hunk 26]
```diff
@@ -0,0 +1,409 @@
...
+void cmsis_dap_usb_process(k_timeout_t timeout)
+{
+	struct net_buf *buf;
+	size_t len;
+	int ret;
+
+	buf = k_fifo_get(&ep_out_queue, timeout);
+
+	len = dap_execute_cmd(buf->data, tx_buf);
+	LOG_DBG("response length %u, starting with [0x%02X, 0x%02X]", len, tx_buf[0], tx_buf[1]);
```
Comment (jfischer-no) [inline samples/subsys/dap/src/cmsis_dap_usb.c]:
```
	LOG_DBG("response length %u, starting with [0x%02X, 0x%02X]",
		len, tx_buf[0], tx_buf[1]);
```

File: drivers/dp/Kconfig [hunk 27]
```diff
@@ -24,7 +24,18 @@ config SWDP_BITBANG_DRIVER
...
 	help
 	  Serial Wire Debug Port bit-bang driver.
 
+if SWDP_BITBANG_DRIVER
+
+config SWDP_BITBANG_DRIVER_SEPARATE_IN_OUT
+	bool "SWDIO is connected through transceiver"
+	default n
+	help
+	  Instead of DIO, DIN, DOUT and DNOE should be defined.
```
Comment (jfischer-no) [inline drivers/dp/Kconfig:37]: Driver supports multiples instances, therefore this cannot be a Kconfig option.

File: drivers/dp/fast_bitbang.h [hunk 28]
```diff
@@ -0,0 +1,105 @@
...
+static ALWAYS_INLINE void pin_platform_set(void *base, uint32_t port, uint8_t pin)
+{
+#if defined(CONFIG_SOC_SERIES_NRF52X) || defined(CONFIG_SOC_SERIES_NRF53X)
+	NRF_GPIO_Type * reg = base;
+
+	reg->OUTSET = BIT(pin);
+#elif defined(CONFIG_SOC_FAMILY_LPC)
+	GPIO_Type *gpio_base = base;
+
+	gpio_base->SET[port] = BIT(pin);
```
Comment (jfischer-no) [inline drivers/dp/fast_bitbang.h]: This is still wrong because gpio_base->SET[port] can be passed as base, there is no reason for port parameter.

File: drivers/dp/fast_bitbang.h [hunk 29]
```diff
@@ -0,0 +1,105 @@
...
+#if defined(CONFIG_SOC_SERIES_NRF52X) || defined(CONFIG_SOC_SERIES_NRF53X)
+	NRF_GPIO_Type * reg = base;
+
+	reg->PIN_CNF[pin] = 0b0001;
+#elif defined(CONFIG_SOC_FAMILY_LPC)
+	GPIO_Type *gpio_base = base;
+
+	gpio_base->DIRSET[port] = BIT(pin);
+#endif
+}
```
Comment (jfischer-no) [inline drivers/dp/fast_bitbang.h]: Support for the single IN/OUT pin should be a separate commit.

File: drivers/dp/swdp_bitbang.c [hunk 30]
```diff
@@ -20,15 +20,11 @@
 #include <zephyr/drivers/gpio.h>
 #include <zephyr/drivers/swdp.h>
 
+#include "fast_bitbang.h"
```
Comment (jfischer-no) [inline drivers/dp/swdp_bitbang.c]: #include "swdp_ll_bitbang.h"

File: drivers/dp/fast_bitbang.h [hunk 31]
```diff
@@ -0,0 +1,105 @@
...
+
+
+static ALWAYS_INLINE void pin_platform_set(void *base, uint32_t port, uint8_t pin)
```
Comment (jfischer-no) [inline drivers/dp/fast_bitbang.h]: `static inline void swdp_ll_pin_set(void *const base, const uint8_t pin)`

File: drivers/dp/fast_bitbang.h [hunk 32]
```diff
@@ -0,0 +1,105 @@
...
+	NRF_GPIO_Type * reg = base;
+
+	reg->PIN_CNF[pin] = 0b0000;
+#elif defined(CONFIG_SOC_FAMILY_LPC)
+	GPIO_Type *gpio_base = base;
+	gpio_base->DIRCLR[port] = BIT(pin);
+#endif
+}
+
+static ALWAYS_INLINE void pin_platform_config_output(void *base, uint32_t port, uint8_t pin)
```
Comment (jfischer-no) [inline drivers/dp/fast_bitbang.h]: `static inline void swdp_ll_pin_output(void *cont base, const uint8_t pin)`

File: drivers/dp/swdp_bitbang.c [hunk 33]
```diff
@@ -101,124 +103,108 @@ static ALWAYS_INLINE uint32_t sw_get32bit_parity(uint32_t data)
...
 {
 	const struct sw_config *config = dev->config;
-	const struct gpio_dt_spec *dt_spec = &config->clk;
 
-	pin_platform_set(config->clk_reg, config->clk_port, dt_spec->pin);
+#if FAST_BITBANG_HW_SUPPORT
+	pin_platform_set(config->clk_reg, config->clk_port, config->clk.pin);
+#else
+	gpio_pin_set_dt(&config->clk, 1);
+#endif /* FAST_BITBANG_HW_SUPPORT */
```
Comment (jfischer-no) [inline drivers/dp/swdp_bitbang.c]: Please not more ifdefery.
```c
	if (FAST_BITBANG_HW_SUPPORT) {
		swdp_ll_pin_set(config->clk_reg, config->clk.pin);
	} else {
		gpio_pin_set_dt(&config->clk, 1);
	}
```

**Review by jfischer-no — CHANGES_REQUESTED** [review]

File: subsys/dap/Kconfig (hunk 1 above)
Comment (jfischer-no) [reply subsys/dap/Kconfig:21]: This Kconfig option is not exclusive for USB transport. You have to adapt it according to your transport layer needs.

File: samples/subsys/dap/src/cmsis_dap_usb.c (hunk 18 above)
Comment (maxd-nordic) [reply samples/subsys/dap/src/cmsis_dap_usb.c]: If we have multiple interfaces, we need this function subset header. If we don't, we cannot have the header. I don't know of a better hint on whether we have multiple interfaces.

File: drivers/dp/Kconfig (hunk 27 above)
Comment (maxd-nordic) [reply drivers/dp/Kconfig:37]: Could you show me a way how to implement this only using device tree?
Also, I'm not sure if I see a use case where you'd have one instance with and another instance without separate I/O lines.

File: drivers/dp/fast_bitbang.h (hunk 28 above)
Comment (maxd-nordic) [reply drivers/dp/fast_bitbang.h]: Currently, `SW_GPIOS_GET_REG` is used to get the register, which points to `gpio_base`. If we wanted to pass the `SET[port]` register directly, we would have to do chip-specific things in `SW_DEVICE_DEFINE`. I would like to avoid that.

File: drivers/dp/swdp_bitbang.c (hunk 33 above)
Comment (maxd-nordic) [reply drivers/dp/swdp_bitbang.c]: What's the alternative?

File: subsys/dap/cmsis_dap.c (hunk 3 above)
Comment (maxd-nordic) [reply subsys/dap/cmsis_dap.c]: The output of `sizeof` is `size_t` and would cause a compiler warning if not casted explicitly.

File: subsys/dap/cmsis_dap.c (hunk 6 above)
Comment (maxd-nordic) [reply subsys/dap/cmsis_dap.c:679]: This implements the new SWDP_SEQUENCE required by the current protocol.

File: samples/subsys/dap/boards/nrf52840dk_nrf52840.overlay (hunk 16 above)
Comment (maxd-nordic) [reply samples/subsys/dap/boards/nrf52840dk_nrf52840.overlay]: that was only for testing, I'll remove it

File: subsys/dap/Kconfig (hunk 1 above)
Comment (maxd-nordic) [reply subsys/dap/Kconfig:21]: Right, the idea was to potentially have DAP over BLE at some point.

File: subsys/dap/cmsis_dap.c (hunk 4 above)
Comment (maxd-nordic) [reply subsys/dap/cmsis_dap.c]: note to self: the limit is going to be `CMSIS_DAP_PACKET_SIZE - 2`

**Comment by maxd-nordic** [comment]
@jfischer-no thanks for the review! I'll try to fulfill the requests, but it might take a while to separate the commits.

File: drivers/dp/swdp_bitbang.c (hunk 33 above)
Comment (tmon-nordic) [reply drivers/dp/swdp_bitbang.c]: The main thing is that if you do `if (FAST_BITBANG_HW_SUPPORT) {` as proposed by @jfischer-no instead of `#if FAST_BITBANG_HW_SUPPORT` then the code will fail to compile if there is error regardless of the `FAST_BITBANG_HW_SUPPORT` value. With ifdef, the path that is not in the default configuration is likely to get completely broken over time when the other parts of code change, but the inactive ifdef is left intact.

Note that the compiler will remove the path not taken and thus the resulting binary will be the same in both cases.

File: samples/subsys/dap/boards/nrf52840dk_nrf52840.conf (hunk 15 above)
Comment (maxd-nordic) [reply samples/subsys/dap/boards/nrf52840dk_nrf52840.conf]: removed

File: samples/subsys/dap/src/cmsis_dap_usb.c (hunk 17 above)
Comment (maxd-nordic) [reply samples/subsys/dap/src/cmsis_dap_usb.c]: will be fixed once this is merged: https://github.com/zephyrproject-rtos/zephyr/pull/59758

File: samples/subsys/dap/src/cmsis_dap_usb.c (hunk 18 above)
Comment (maxd-nordic) [reply samples/subsys/dap/src/cmsis_dap_usb.c]: I guess for my use case (using it as part of connectivity-bridge), this file would actually be a library.

File: samples/subsys/dap/src/cmsis_dap_usb.c (hunk 22 above)
Comment (maxd-nordic) [reply samples/subsys/dap/src/cmsis_dap_usb.c]: It makes sense to me from a USB perspective, but I'll change it.

File: samples/subsys/dap/src/cmsis_dap_usb.c (hunk 23 above)
Comment (maxd-nordic) [reply samples/subsys/dap/src/cmsis_dap_usb.c]: I'd like to keep them so it's easy to add interfaces.

File: samples/subsys/dap/src/cmsis_dap_usb.c (hunk 24 above)
Comment (maxd-nordic) [reply samples/subsys/dap/src/cmsis_dap_usb.c]: removed

File: samples/subsys/dap/src/cmsis_dap_usb.c (hunk 25 above)
Comment (maxd-nordic) [reply samples/subsys/dap/src/cmsis_dap_usb.c]: done

File: drivers/dp/fast_bitbang.h (hunk 28 above)
Comment (maxd-nordic) [reply drivers/dp/fast_bitbang.h]: Also, SET[..] is not the only register, we also have a DIRSET[..] and DIRCLR[..] register..

File: drivers/dp/swdp_bitbang.c (hunk 33 above)
Comment (maxd-nordic) [reply drivers/dp/swdp_bitbang.c]: done

File: subsys/dap/cmsis_dap.h (hunk 7 above)
Comment (maxd-nordic) [reply subsys/dap/cmsis_dap.h]: done

File: samples/subsys/dap/CMakeLists.txt (hunk 14 above)
Comment (maxd-nordic) [reply samples/subsys/dap/CMakeLists.txt]: removed

File: subsys/dap/cmsis_dap.h (hunk 11 above)
Comment (maxd-nordic) [reply subsys/dap/cmsis_dap.h]: can these be removed entirely?

File: subsys/dap/Kconfig (hunk 2 above)
Comment (maxd-nordic) [reply subsys/dap/Kconfig:50]: AFAICT, the cmsis_dap.c part only supports one instance. Maybe we can defer that multi-instance support to a later PR entirely.

File: subsys/dap/cmsis_dap.c (hunk 4 above)
Comment (maxd-nordic) [reply subsys/dap/cmsis_dap.c]: checks have been added

**Comment by maxd-nordic** [comment]
@jfischer-no @tmon-nordic please have another look. I tended to the request where I had an idea to fix. Might need more input if more things need to be fixed.

File: subsys/dap/cmsis_dap.c (hunk 3 above)
Comment (jfischer-no) [reply subsys/dap/cmsis_dap.c]: Certainly not. Please remove all the unnecessary (uint8_t).
btw, 
```c
		memcpy(info, CONFIG_CMSIS_DAP_PROBE_VENDOR,
		       sizeof(CONFIG_CMSIS_DAP_PROBE_VENDOR));
```

File: samples/subsys/dap/src/cmsis_dap_usb.c (hunk 17 above)
Comment (jfischer-no) [reply samples/subsys/dap/src/cmsis_dap_usb.c]: `_t`  are still there.

File: samples/subsys/dap/src/cmsis_dap_usb.c (hunk 23 above)
Comment (jfischer-no) [reply samples/subsys/dap/src/cmsis_dap_usb.c]: I do not accept AUTO_EP_OUT/AUTO_EP_IN macros.

File: drivers/dp/Kconfig (hunk 27 above)
Comment (jfischer-no) [reply drivers/dp/Kconfig:37]: > Also, I'm not sure if I see a use case where you'd have one instance with and another instance without separate I/O lines.

Well it is now we design the drivers, and Kconfig is not correct approach for this option. I can fix it for you.

File: drivers/dp/fast_bitbang.h (hunk 28 above)
Comment (jfischer-no) [reply drivers/dp/fast_bitbang.h]: Yes and we do not need port parameter for that, please drop this "LPC" patch and the relevant changes in your patches.

File: drivers/dp/Kconfig (hunk 27 above)
Comment (maxd-nordic) [reply drivers/dp/Kconfig:37]: Please do.

File: samples/subsys/dap/src/cmsis_dap_usb.c (hunk 23 above)
Comment (maxd-nordic) [reply samples/subsys/dap/src/cmsis_dap_usb.c]: done

File: samples/subsys/dap/src/cmsis_dap_usb.c (hunk 17 above)
Comment (maxd-nordic) [reply samples/subsys/dap/src/cmsis_dap_usb.c]: done

File: subsys/dap/cmsis_dap.c (hunk 3 above)
Comment (maxd-nordic) [reply subsys/dap/cmsis_dap.c]: done

File: drivers/dp/fast_bitbang.h (hunk 28 above)
Comment (maxd-nordic) [reply drivers/dp/fast_bitbang.h]: done

**Comment by jfischer-no** [comment]
Applied to upstream PR with significant changes.
